src/shared/test/bridge-test.js

```diff
@@ -10,246 +10,476 @@ describe('shared/bridge', () => {
   beforeEach(() => {
```

Comment:
The existing `Bridge` tests are unit tests that stub `RPC.call`. The new tests are integration tests that end up testing a lot of `RPC`-specific details. We generally try to avoid mixing unit and integration tests in the same module, as becomes unclear for developers making changes later which approach they should follow when adding a new test. Sometimes there might be cases where there is justification to have a test that deviates from the other tests in terms of its unit-ness. In that case it is good to call out the fact that these are an exception. For example with a specific "integration tests" `describe` block.

At the time when the `Bridge` tests were originally written I don't think we had a good mocking tool (`$imports.$mock`) which is probably why they create a real `RPC` and then stub `call` each time. This could be simplified to use `$imports.$mock` now. eg:

```js
// Active channels
let channels = [];

class FakeRPC {
  constructor(sourceFrame, ...) {
    this.sourceFrame = sourceFrame;
    ...
    
    this.call = sinon.stub().resolves(...)
    channels.push(this);
  }
  
  destroy() {
    channels.splice(channels.indexOf(this), 1);
  }
}

...
$imports.$mock({ './frame-rpc': { RPC: FakeRPC }})
```

I think we are eventually going to want some integration tests for inter-frame messaging. Following the examples of `src/{annotator, sidebar}/test/integration` these could go in `src/shared/test/integration/messaging-test.js`.

Reply:
It is good to know about the integration tests.

Reply:
Until this [PR](https://github.com/hypothesis/client/pull/3565), `FrameRPC` didn't have unit testing. I think the `Bridge` was supposed to unit test `FrameRPC`, however, because of the difficulty of creating two frames to communicate with each other, most of the methods in `FrameRPC` were mocked.